File: LetCode/3092. Most Frequent IDs/solution.cpp
#include <vector>
#include <unordered_map>
using namespace std;

/*
Runtime
193
ms
Beats
96.08%
of users with C++
Memory
136.27
MB
Beats
98.73%
of users with C++
*/

class Solution {
public:
    vector<long long> mostFrequentIDs(vector<int>& nums, vector<int>& freq) {
        int n = nums.size();
        vector<long long> ans(n, 0);
        unordered_map<int, long long> mp;

        ans[0] = freq[0];
        mp[nums[0]] = freq[0];
        int maxNum = nums[0];

        for(int i = 1; i < n; i++){

            mp[nums[i]] += freq[i];

            if(maxNum == nums[i] && freq[i] < 0){
                // find new maximum?
                long long currMax = mp[maxNum];
                for(auto p : mp){
                    if(p.second > currMax){
                        currMax = p.second;
                        maxNum = p.first;
                    }
                }
            }else if(mp[nums[i]] > mp[maxNum]){
                maxNum = nums[i];
            }
            ans[i] = mp[maxNum];
        }

        return ans;
    }
};

File: LetCode/3092. Most Frequent IDs/solution.cs

/*
Runtime
497
ms
Beats
85.00%
of users with C#
Memory
93.94
MB
Beats
86.25%
of users with C#
*/
public class Solution {
    public long[] MostFrequentIDs(int[] nums, int[] freq) {
        int n = nums.Length;
        long[] ans = new long[n];
        
        Dictionary<int, long> dict = new();

        dict.Add(nums[0], freq[0]);
        ans[0] = freq[0];
        int maxNum = nums[0];
        
        for(int i = 1; i < n; i++){
            
            if(dict.ContainsKey(nums[i])){
                dict[nums[i]] += freq[i];
            }else{
                dict.Add(nums[i], freq[i]);
            }
            
            if(maxNum == nums[i] && freq[i] < 0){
                long maxFreq = dict[maxNum];
                foreach(var pair in dict){
                    if(pair.Value > maxFreq){
                        maxFreq = pair.Value;
                        maxNum = pair.Key;
                    }
                }
            }else if(dict[nums[i]] > dict[maxNum]){
                maxNum = nums[i];
            }
            
            
            ans[i] = dict[maxNum];
        }
        
        return ans;
    }
}




// Time Limit Exceeded
public class Solution {
    public long[] MostFrequentIDs(int[] nums, int[] freq) {
        int n = nums.Length;
        long[] ans = new long[n];
        
        Dictionary<int, int> dict = new();
        
        for(int i = 0; i < n; i++){
            
            if(dict.ContainsKey(nums[i])){
                dict[nums[i]] += freq[i];
            }else{
                dict.Add(nums[i], freq[i]);
            }
            
            int maxFreq = 0;
            foreach(var pair in dict){
                if(pair.Value > maxFreq)
                    maxFreq = pair.Value;
            }
            
            ans[i] = maxFreq;
            
        }
        
        return ans;
    }
}